### DISCUSSION TITLE
Rectx documentation generation

### DISCUSSION THREAD

**Discussion by hrszpuk** [hrszpuk/rectx#39] [Ideas]
## What is documentation generation?
It's a programming tool that generates documentation about a project intended for programmers, end users, or both.
The generator would scan the source files of a project and generate documentation in various formats. 
These types of generators are helpful for creating documentation for large projects, or for teams that do not have enough time to write documentation for their project.

## Why does ReCT need it?
Many mainstream, and non-mainstream, languages have documentation generation tools/frameworks (examples: Doxygen, javadoc) that make it easy for anyone to create documentation about a project.

## What would it look like?
This would be a built-in tool apart of rectx, but it would remain opt-in. A specific directory would be created to hold documentation (i.e. `Docs/`) that would be filled with documentation after the command `rectx docgen` is called.

Documentation would be generated from the .rct source files present in the source directory. Source files would contain "doc comments" which would symbolise the documentation different structures in the source files.

The formats I was thinking of supporting would be text, markdown, and HTML.

### Purposed Syntax
```
/// <-- Single line doc comment
/** <-- Multi line doc comment
```

### Examples

**Simple text docs**
ReCT code:
```js
package sys;

/**
 Absolutely amazing Dog class that can woof!!!
*/
class Dog {
    /// Eggs beans bacon
    function woof() {
       sys::Printf("Woof!");
    }
}
```
Generated documentation:
```
Class Dog
Absolutely amazing Dog class that can woof!!!

Functions
woof()
Eggs beans bacon
```

**Markdown example**
ReCT code:
```js
package sys;

/**
The Person class represents a real life person in my game.
This class has a name and an age.
The age can be increased and you can also kill the person using the Die function.*/
class Person {
    set string name;
    set int age; /// Amazing comment
    
    function Constructor() {
        name <- "Jimmy";
        age <- 17;
    }

    /// This function kills the person so be very careful with it!
    set function Die() {
        sys::Printf("Uh oh!");
    }

    /**
    This function increments the age of the Person object.
    Once the person exceeds the age of 100 the Die function is called.
    */
    set function GrowUp() {
        age++;
        if (age = 100) {
            Die();
        }
    }
}
```
Markdown:
```md
# Class Person
The Person class represents a real life person in my game.
This class has a name and an age.
The age can be increased and you can also kill the person using the Die function.

## Attributes
**name string**
**age int - Amazing comment**

## Methods

### Die()
This function kills the person so be very careful with it!

### GrowUp()
This function increments the age of the Person object.
Once the person exceeds the age of 100 the Die function is called.
```

**Comment by hrszpuk** [comment]
Open to suggestions and critiques about the idea.

**Comment by Khhs167** [comment]
JavaDoc, pdoc, Doxygen. Every good language has it. Wonder if doxygen could support ReCT actually?

**Reply by mr-cheffy** [reply]
ig that if rect has the same comments, maybe it could

**Comment by mr-cheffy** [comment]
What about function information?

`@arg name <type> - desc`
`@return <type>`
`@desc desc`
`@short short desc`

etc... or is that already "auto-implemented" by some sort of parser?

**Reply by hrszpuk** [reply]
I think it would be easier for the users if it was just a parser. So instead of using the `@` statements like in your example they just write rect and the syntax is taken and used in the documentation. So, function arguments would just be taken from the function declaration instead of the user specifying them specifically.
I like the idea of having short and long descriptions for functions and classes though.

**Reply by mr-cheffy** [reply]
Maybe, we can still have the arguments as a backup, just in case

**Comment by mr-cheffy** [comment]
We would need to access the Rect parser to do this and for it to be "up to date"
